### PR TITLE
Vue comp checkbox enable field

### PR DESCRIPTION
**Pull request by bjorvack**
## Type

- Enhancement

## Pull request description

Using VueJs to handle the checkbox to enable the linked field.

### CODE REVIEW

File: src/Backend/Modules/Pages/Resources/views/FormLayout.html.twig [hunk 1]
```diff
@@ -139,52 +135,80 @@
...
+          ref="{{ form.description.vars.id }}"
+          checkbox-id="{{ form.descriptionOverwrite.vars.id }}"
+          checkbox-label="{{ form.descriptionOverwrite.vars.label|trans|capitalize }}"
+          checkbox-name="{{ form.descriptionOverwrite.vars.full_name }}"
+          field-id="{{ form.description.vars.id }}"
+          field-label="{{  form.description.vars.label|trans|capitalize }}"
+          field-name="{{ form.description.vars.full_name }}"
+          field-help-text="{{  'msg.HelpMetaDescription'|trans|raw }}"
+          value-on-load="{{ form.description.vars.value }}"
+          {% if form.descriptionOverwrite.vars.checked %} enabled-on-load {% endif %}
```
Comment (Katrienvh) [inline src/Backend/Modules/Pages/Resources/views/FormLayout.html.twig:190]: whitespace after `enabled-on-load ` to much, this issue appears multiply time in this file and the other files where the same code was added

**Review by Katrienvh — CHANGES_REQUESTED** [review]

File: src/Backend/Modules/Pages/Resources/views/FormLayout.html.twig (hunk 1 above)
Comment (bjorvack) [reply src/Backend/Modules/Pages/Resources/views/FormLayout.html.twig]: fixed ae55c25edd5b331110f8a54c002734192f656ca1

**Review by Katrienvh — APPROVED** [review]

File: package-lock.json [hunk 2]
```diff
@@ -1116,9 +1116,9 @@
       }
     },
     "@editorjs/editorjs": {
-      "version": "2.19.0",
-      "resolved": "https://registry.npmjs.org/@editorjs/editorjs/-/editorjs-2.19.0.tgz",
-      "integrity": "sha512-8PUVaBZx69IrG8dNrE+FZbHSiRTR8ql8L/cmEi1mOdEdTqnOLq5Wv9dgemK00mBWEgNoavMAjtGQpItGknAa8A==",
+      "version": "2.18.0",
+      "resolved": "https://registry.npmjs.org/@editorjs/editorjs/-/editorjs-2.18.0.tgz",
+      "integrity": "sha512-9OKs580JFKoXCAw7llb19E+qxY6QuzgDBq50cKbyOS1Lt+BglTq/zBdXxmRWNRTlCMxjTB1vgnq70+OjEyDSlw==",
```
Comment (carakas) [inline package-lock.json:1121]: why downgrading the editor?

File: package.json [hunk 3]
```diff
@@ -7,7 +7,7 @@
   "author": "Tijs Verkoyen <tijs@verkoyen.eu>",
   "license": "MIT",
   "devDependencies": {
-    "@editorjs/editorjs": "^2.19.0",
+    "@editorjs/editorjs": "^2.16.1",
```
Comment (carakas) [inline package.json:10]: why .16 instead of .19?

File: src/Backend/Form/Type/CheckboxEnableField.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+namespace Backend\Form\Type;
+
+use Backend\Form\DataTransferObject\EnabledFieldDataTransferObject;
+use Common\Form\SwitchType;
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class CheckboxEnableField extends AbstractType
```
Comment (carakas) [inline src/Backend/Form/Type/CheckboxEnableField.php:12]: your DTO is called EnabledField and your form type is called CheckboxEnableField
Don't like this naming inconsistency.
Plus all form types need to have the suffix Type

File: src/Backend/Modules/Blog/Layout/Templates/Edit.html.twig [hunk 5]
```diff
@@ -17,6 +17,19 @@
 {% endblock %}
 
 {% block content %}
+  <div id="app">
```
Comment (carakas) [inline src/Backend/Modules/Blog/Layout/Templates/Edit.html.twig:20]: I assume this is leftover test data?

**Review by carakas — CHANGES_REQUESTED** [review]
Did you also add the new checkbox type to the SpoonForm SEO form? Or is that not needed?

File: src/Backend/Form/Type/CheckboxEnableField.php (hunk 4 above)
Comment (bjorvack) [reply src/Backend/Form/Type/CheckboxEnableField.php]: fixed 42f2d5c18ecef0a634f028a335a02722b02c9528

File: src/Backend/Modules/Blog/Layout/Templates/Edit.html.twig (hunk 5 above)
Comment (bjorvack) [reply src/Backend/Modules/Blog/Layout/Templates/Edit.html.twig:20]: No, vue needs an id to define where it should process the components

File: package.json (hunk 3 above)
Comment (bjorvack) [reply package.json]: fixed in 9d4b206

File: package-lock.json (hunk 2 above)
Comment (bjorvack) [reply package-lock.json]: fixed in 9d4b206

File: package.json [hunk 6]
```diff
@@ -50,7 +50,7 @@
     "typeahead.js": "^0.11.1",
     "url-loader": "^4.1.0",
     "vue": "^2.6.12",
-    "vue-loader": "^15.9.6",
```
Comment (carakas) [inline package.json:53]: why did you lower the version?

File: src/Backend/Core/Js/BlockEditor/BlockEditor.js [hunk 7]
```diff
@@ -5,10 +5,10 @@ import List from '@editorjs/list'
 import Paragraph from '@editorjs/paragraph'
 import Underline from '@editorjs/underline'
 import MediaLibraryImage from './Blocks/MediaLibraryImage'
-import MediaLibraryVideo from './Blocks/MediaLibraryVideo'
 import Button from './Blocks/Button'
 import Quote from './Blocks/Quote'
 import Raw from './Blocks/Raw'
+import MediaLibraryVideo from './Blocks/MediaLibraryVideo'
```
Comment (carakas) [inline src/Backend/Core/Js/BlockEditor/BlockEditor.js:11]: why did you move this?

File: src/Backend/Core/Js/Components/Meta.js [hunk 8]
```diff
@@ -109,18 +81,23 @@ export class Meta {
       // calculate meta
       function calculateMeta (e, element) {
         const title = (typeof element !== 'undefined') ? element.val() : $(this).val()
+        const seoVueRefs = window.backend.seoVue.$refs
 
-        if ($pageTitle.length > 0 && $pageTitleOverwrite.length > 0) {
-          if (!$pageTitleOverwrite.is(':checked')) $pageTitle.val(title)
+        if (seoVueRefs[$pageTitle] !== undefined) {
```
Comment (carakas) [inline src/Backend/Core/Js/Components/Meta.js:86]: plsease use `typeof seoVueRefs[$pageTitle] !== 'undefined'` instead, this is not a safe way of checking it

File: src/Backend/Core/Js/Components/Meta.js [hunk 9]
```diff
@@ -109,18 +81,23 @@ export class Meta {
...
+          seoVueRefs[$pageTitle].updateValue(title)
         }
 
-        if ($navigationTitle.length > 0 && $navigationTitleOverwrite.length > 0) {
-          if (!$navigationTitleOverwrite.is(':checked')) $navigationTitle.val(title)
+        if (seoVueRefs[$navigationTitle] !== undefined) {
```
Comment (carakas) [inline src/Backend/Core/Js/Components/Meta.js:90]: plsease use `typeof seoVueRefs[$navigationTitle] !== 'undefined'` instead, this is not a safe way of checking it

File: src/Backend/Core/Js/Components/Meta.js [hunk 10]
```diff
@@ -109,18 +81,23 @@ export class Meta {
...
+          seoVueRefs[$navigationTitle].updateValue(title)
         }
 
-        if (!$metaDescriptionOverwrite.is(':checked')) $metaDescription.val(title)
+        if (seoVueRefs[$metaDescription] !== undefined) {
```
Comment (carakas) [inline src/Backend/Core/Js/Components/Meta.js:94]: plsease use `typeof seoVueRefs[$metaDescription] !== 'undefined'` instead, this is not a safe way of checking it

File: src/Backend/Core/Js/Components/Meta.js [hunk 11]
```diff
@@ -109,18 +81,23 @@ export class Meta {
...
+          seoVueRefs[$metaDescription].updateValue(title)
+        }
 
-        if (!$metaKeywordsOverwrite.is(':checked')) $metaKeywords.val(title)
+        if (seoVueRefs[$metaKeywords] !== undefined) {
```
Comment (carakas) [inline src/Backend/Core/Js/Components/Meta.js:98]: plsease use `typeof seoVueRefs[$metaKeywords] !== 'undefined'` instead, this is not a safe way of checking it

File: src/Backend/Core/Layout/Templates/FormLayout.html.twig [hunk 12]
```diff
@@ -118,41 +129,59 @@
...
-        <div class="form-group">
-          {{ form_widget(form.titleOverwrite) }}
-        </div>
-        <div class="form-group">
-          {{ form_label(form.title) }}
-          {{ form_widget(form.title) }}
-          {{ form_errors(form.title) }}
-          <small class="form-text text-muted">{{ 'msg.HelpPageTitle'|trans|raw }}</small>
-        </div>
+        <checkbox-enable-field
```
Comment (carakas) [inline src/Backend/Core/Layout/Templates/FormLayout.html.twig:132]: seems like a lot of repeating config, can you create a macro for this so that it will become reusable?

File: src/Backend/Form/DataTransferObject/CheckboxEnabledFieldDataTransferObject.php [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+class CheckboxEnabledFieldDataTransferObject
+{
+    public bool $enableField = false;
+
+    public string $field = '';
+
+    public static function create(bool $enableField, string $field): self
+    {
+        $dto = new self();
```
Comment (carakas) [inline src/Backend/Form/DataTransferObject/CheckboxEnabledFieldDataTransferObject.php:13]: not a fan of this syntax, either create a private constructor or just use the constructor

File: src/Backend/Form/Type/CheckboxEnableFieldType.php [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+                'enableField',
+                SwitchType::class,
+                [
+                    'required' => false,
+                    'label' => $options['enable_field_label']
+                ]
+            )
+            ->add(
+                'field',
+                TextType::class,
```
Comment (carakas) [inline src/Backend/Form/Type/CheckboxEnableFieldType.php:27]: I would make the type of field configurable and the options for it as well, that way it is way more reusable a little bit like how you can configure a CollectionType

File: src/Backend/Form/DataTransferObject/CheckboxEnabledFieldDataTransferObject.php [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Backend\Form\DataTransferObject;
+
+class CheckboxEnabledFieldDataTransferObject
+{
+    public bool $enableField = false;
+
+    public string $field = '';
+
+    public static function create(bool $enableField, string $field): self
```
Comment (carakas) [inline src/Backend/Form/DataTransferObject/CheckboxEnabledFieldDataTransferObject.php:11]: Not a fan of the naming here, field makes me think it is the name of the form field instead of the value.
I would suggest using `enabled` and `value` instead.
Also I would drop the string type and changed it to mixed so that we can use it for datefields etc as well

File: src/Backend/Modules/Blog/Layout/Templates/Edit.html.twig (hunk 5 above)
Comment (carakas) [reply src/Backend/Modules/Blog/Layout/Templates/Edit.html.twig:20]: @bjorvack your id is `checkbox-id="1234-check"` ...
This is a random checkbox in the edit page of the blog so can you give me a reason why this isn't test data?

**Review by carakas — CHANGES_REQUESTED** [review]

File: src/Backend/Form/Type/CheckboxEnableFieldType.php (hunk 14 above)
Comment (tijsverkoyen) [reply src/Backend/Form/Type/CheckboxEnableFieldType.php:27]: @carakas I would do this in a later stage. This is just the initial start of using more Vue components. With this we can see if this works in real projects and improve it based on the actual feedback.

**Comment by carakas** [comment]
@bjorvack do you know how to resolve the merge conflict?
